File: src/main/PartitionEqualSubsetSum.java
import java.util.Arrays;

import static util.ArrayUtils.array;
import static util.ResourceUtils.loadTestcase;

/**
 * 416. 分割等和子集：https://leetcode.cn/problems/partition-equal-subset-sum/
 */
public class PartitionEqualSubsetSum {

    public static void main(String[] args) {
        System.out.println(new PartitionEqualSubsetSum().canPartition(new int[] {1,5,11,5})); // true
        System.out.println(new PartitionEqualSubsetSum().canPartition(new int[] {1,2,3,5})); // false
        System.out.println(new PartitionEqualSubsetSum().canPartition(new int[] {2,2,3,5})); // false
        System.out.println(new PartitionEqualSubsetSum().canPartition(array(
            loadTestcase("416-testcase-1.txt")))); // false
    }

    public boolean canPartition(int[] nums) {
        int n = nums.length;
        if (n < 2) {
            return false;
        }
        int sum = Arrays.stream(nums).sum(), maxNum = Arrays.stream(nums).max().getAsInt();
        if (sum % 2 != 0) {
            return false;
        }
        int target = sum / 2;
        if (maxNum > target) {
            return false;
        }
        boolean[] dp = new boolean[target + 1];
        dp[0] = true;
        for (int i = 0; i < n; i++) {
            for (int j = target; j >= nums[i]; j--) {
                dp[j] |= dp[j - nums[i]];
            }
        }
        return dp[target];
    }

}

File: c/partition-equal-subset-sum.c
/**
 * 416. 分割等和子集：https://leetcode.cn/problems/partition-equal-subset-sum
 */

#include <stdbool.h>
#include <stdlib.h>

#define MIN(a, b) ((b) < (a) ? (b) : (a))

bool canPartition(int* nums, int n) {
    int sum = 0;
    for (int i = 0; i < n; i++) {
        sum += nums[i];
    }
    if (sum & 1) {
        return false;
    }
    int k = sum >> 1;
    bool * dp = calloc(k + 1, sizeof(bool));
    dp[0] = true;
    int s = 0;
    for (int i = 0; i < n; i++) {
        int num = nums[i];
        s = MIN(s + num, k);
        for (int j = s; j >= num; j--) {
            dp[j] |= dp[j - num];
        }
        if (dp[k]) {
            free(dp);
            return true;
        }
    }
    free(dp);
    return false;
}
